File: chapter11/typescript/Desktop.ts
import { Computer } from "./Computer";
import { OperatingSystem } from "./OperatingSystem";

/**
 * Desktop extends Computer (the bridge "abstraction") and delegates all of the real work to OperatingSystem Implementor
 *
 * @Author: Bridget
 */
 export class Desktop extends Computer {

    /**
     * call super to inheritance os instance
     *
     * @param operatingSystems
     */
    public constructor(operatingSystems: OperatingSystem[]) {
        super(operatingSystems);
    }

    /**
     * call OperatingSystem to display Operating System name, version, memory
     */
    public displayOSInfo(): void {
        console.log("... Desktop Operating System information ...");
        for(let os of this.operatingSystems) {
            os.displayOSName();
            os.displayOSVersion();
            os.displayMemory();
        }

    }

    /**
     * start Desktop Operating System
     */
    public start(): void {
        console.log("... Desktop Operating System starting ...");
        for(let os of this.operatingSystems) {
            os.start();
        }
    }
    /**
     * shutdown Desktop Operating System
     */
    public shutdown(): void {
        console.log("... Desktop Operating System shutting down ...");
        for(let os of this.operatingSystems) {
            os.shutdown();
        }
    }
    /**
     * restart Desktop Operating System
     */
    public restart(): void {
        console.log("... Desktop Operating System restarting ...");
        for(let os of this.operatingSystems) {
            os.shutdown();
            os.start();
        }
    }
}

File: chapter11/java/designpattern/src/main/java/com/designpattern/Desktop.java
package com.designpattern;

import java.util.List;

/**
 * Desktop extends Computer (the bridge "abstraction") and delegates all of the real work to OperatingSystem Implementor
 *
 * @version 1.0.0
 * @Author Bridget
 * @license MIT
 */
public class Desktop extends Computer {

    /**
     * call super to inheritance os instance
     *
     * @param operatingSystems
     */
    public Desktop(List<OperatingSystem> operatingSystems) {
        super(operatingSystems);
    }

    /**
     * call OperatingSystem to display Operating System name, version, memory
     */
    @Override
    public void displayOSInfo() {
        System.out.println("... Desktop Operating System information ...");
        for (OperatingSystem os : operatingSystems) {
            os.displayOSName();
            os.displayOSVersion();
            os.displayMemory();
        }

    }

    /**
     * start Desktop Operating System
     */
    @Override
    public void start() {
        System.out.println("... Desktop Operating System starting ...");
        for (OperatingSystem os : operatingSystems) {
            os.start();
        }

    }

    /**
     * shutdown Desktop Operating System
     */
    @Override
    public void shutdown() {
        System.out.println("... Desktop Operating System shutting down ...");
        for (OperatingSystem os : operatingSystems) {
            os.shutdown();
        }
    }

    /**
     * restart Desktop Operating System
     */
    @Override
    public void restart() {
        System.out.println("... Desktop Operating System restarting ...");
        for (OperatingSystem os : operatingSystems) {
            os.shutdown();
            os.start();
        }

    }
}
